### PR TITLE
Fix Matrix4x4.CreateReflection when D is not zero

### LINKED ISSUE #110050: Matrix4x4.CreateReflection gives incorrect results in .NET 9
### Description

In .NET 9, using Matrix4x4.CreateReflection with a plane where D != 0 no longer gives correct results, specifically in the last column of the matrix (M14, M24, M34, M44).

### Reproduction Steps

```csharp
using System;
using System.Numerics;

var plane = new Plane(0, 1, 0, 60);
var mtx = Matrix4x4.CreateReflection(plane);
Console.Out.WriteLine(mtx);
```

### Expected behavior

`{ {M11:1 M12:0 M13:0 M14:0} {M21:0 M22:-1 M23:0 M24:0} {M31:0 M32:0 M33:1 M34:0} {M41:-0 M42:-120 M43:-0 M44:1} }`
(net8)

### Actual behavior

`{ {M11:1 M12:0 M13:0 M14:0} {M21:0 M22:-1 M23:0 M24:-120} {M31:0 M32:0 M33:1 M34:0} {M41:0 M42:-120 M43:0 M44:-7199} }`
(net9)

### Regression?

Yes, the implementation in .NET 8 gave correct results.

### Known Workarounds

a. Take the broken result and reset the values in the last column to (0, 0, 0, 1).
b. Copy the previous implementation to a local project and use that instead.

### Configuration

.NET 9.0.100

### Other information

The implemenation was changed in https://github.com/dotnet/runtime/pull/103527 for .NET 9

Before: <https://github.com/dotnet/runtime/blame/ab013a35ac6d76320bb6e28a49d62ad82a88e5d6/src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.Impl.cs#L720>
After: <https://github.com/dotnet/runtime/blame/3b91ac601980f3cc35e1d8687e7235e874ffc8ea/src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.Impl.cs#L619>

The new implementation accidentally multiples plane.D (W of the vec) by -2 via broadcast whereas the [reference implemenation from DirectXMath linked in the code](https://github.com/microsoft/DirectXMath/blob/0ee92032bcfd509ae72541d9d2f3e8ce4f1bdc4e/Inc/DirectXMathMatrix.inl#L2126) multiplies it by 0

### PR DESCRIPTION
**Pull request by hez2010**
This is a correctness regression from .NET 8 which was introduced in #103527.
Fixed it and added a regression test for it.

Fixes #110050

### CODE REVIEW

File: src/libraries/System.Numerics.Vectors/tests/Matrix4x4Tests.cs [hunk 1]
```diff
@@ -819,11 +819,41 @@ public void Matrix4x4CreateReflectionTest01()
...
+            expected.M32 = 0;
+            expected.M33 = 1;
+            expected.M34 = 0;
+
+            expected.M41 = -0;
+            expected.M42 = -120;
+            expected.M43 = -0;
+            expected.M44 = 1;
+
+            Assert.True(MathHelper.Equal(expected, actual), "Matrix4x4.CreateReflection did not return the expected value.");
```
Comment (tannergooding) [inline src/libraries/System.Numerics.Vectors/tests/Matrix4x4Tests.cs:854]: nit: We should really be updating these tests and ensuring new tests use the better `AssertionExtensions.Equal(float expected, float actual, float allowedVariance)` APIs

In this case, the allowed variance should be `0`.

Comment (hez2010) [reply src/libraries/System.Numerics.Vectors/tests/Matrix4x4Tests.cs:854]: Done.

File: src/libraries/System.Numerics.Vectors/tests/Matrix4x4Tests.cs [hunk 2]
```diff
@@ -819,11 +819,57 @@ public void Matrix4x4CreateReflectionTest01()
                     Vector3 v = point - pp;
                     float d = Vector3.Dot(v, plane.Normal);
                     Vector3 vp = point - 2.0f * d * plane.Normal;
-                    Assert.True(MathHelper.Equal(rp, vp), "Matrix4x4.Reflection did not provide expected value.");
+
+                    Assert.Equal(vp.M11, rp.M11, 0.0f);
```
Comment (tannergooding) [inline src/libraries/System.Numerics.Vectors/tests/Matrix4x4Tests.cs:823]: These should be `AssertExtensions.Equal` so we're using https://github.com/dotnet/runtime/blob/7c1f78fb071ec28a7ba2f9e64efa041e5350215d/src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs#L931, which explicitly handles concepts like `NaN`, `NegativeInfinity`, `NegativeZero`, and other concepts in a way that ensures better determinism/correctness

Comment (hez2010) [reply src/libraries/System.Numerics.Vectors/tests/Matrix4x4Tests.cs]: Done.

File: src/libraries/System.Numerics.Vectors/tests/Matrix4x4Tests.cs [hunk 3]
```diff
@@ -819,11 +819,57 @@ public void Matrix4x4CreateReflectionTest01()
                     Vector3 v = point - pp;
                     float d = Vector3.Dot(v, plane.Normal);
                     Vector3 vp = point - 2.0f * d * plane.Normal;
-                    Assert.True(MathHelper.Equal(rp, vp), "Matrix4x4.Reflection did not provide expected value.");
+
+                    AssertExtensions.Equal(vp.M11, rp.M11, 0.0f);
```
Comment (tannergooding) [inline src/libraries/System.Numerics.Vectors/tests/Matrix4x4Tests.cs:823]: `vp` and `rp` are Vector3, not Matrix4x4

Comment (hez2010) [reply src/libraries/System.Numerics.Vectors/tests/Matrix4x4Tests.cs:823]: You are right. The `Matrix4x4` in the assertion message misled me.

**Comment by hez2010** [comment]
I think we should revert to the `MathHelper.Equal` as the sign of `0.0` in the result is not determined. 

```cs
Plane plane = new Plane(0, 1, 0, 60);
Matrix4x4 actual = Matrix4x4.CreateReflection(plane);

Console.WriteLine(Equal(1.0f, actual.M11));
Console.WriteLine(Equal(0.0f, actual.M12));
Console.WriteLine(Equal(0.0f, actual.M13));
Console.WriteLine(Equal(0.0f, actual.M14));
Console.WriteLine(Equal(0.0f, actual.M21));
Console.WriteLine(Equal(-1.0f, actual.M22));
Console.WriteLine(Equal(0.0f, actual.M23));
Console.WriteLine(Equal(0.0f, actual.M24));
Console.WriteLine(Equal(0.0f, actual.M31));
Console.WriteLine(Equal(0.0f, actual.M32));
Console.WriteLine(Equal(1.0f, actual.M33));
Console.WriteLine(Equal(0.0f, actual.M34));
Console.WriteLine(Equal(-0.0f, actual.M41));
Console.WriteLine(Equal(-120.0f, actual.M42));
Console.WriteLine(Equal(-0.0f, actual.M43));
Console.WriteLine(Equal(1.0f, actual.M44));

(bool, string) Equal(float a, float b)
{
    if (a.ToString() == b.ToString()) return (true, "ok");
    return (false, $"expected: {a}, actual: {b}");
}
```

On .NET 9 with this PR:
```
(True, ok)
(True, ok)
(True, ok)
(True, ok)
(True, ok)
(True, ok)
(True, ok)
(True, ok)
(True, ok)
(True, ok)
(True, ok)
(True, ok)
(False, expected: -0, actual: 0)
(True, ok)
(False, expected: -0, actual: 0)
(True, ok)
```

On .NET 8:
```
(True, ok)
(True, ok)
(True, ok)
(True, ok)
(True, ok)
(True, ok)
(True, ok)
(True, ok)
(True, ok)
(True, ok)
(True, ok)
(True, ok)
(True, ok)
(True, ok)
(True, ok)
(True, ok)
```

On .NET 7:

```
(True, ok)
(False, expected: 0, actual: -0)
(False, expected: 0, actual: -0)
(True, ok)
(False, expected: 0, actual: -0)
(True, ok)
(False, expected: 0, actual: -0)
(True, ok)
(False, expected: 0, actual: -0)
(False, expected: 0, actual: -0)
(True, ok)
(True, ok)
(True, ok)
(True, ok)
(True, ok)
(True, ok)
```

On .NET 6:

```
(True, ok)
(False, expected: 0, actual: -0)
(False, expected: 0, actual: -0)
(True, ok)
(False, expected: 0, actual: -0)
(True, ok)
(False, expected: 0, actual: -0)
(True, ok)
(False, expected: 0, actual: -0)
(False, expected: 0, actual: -0)
(True, ok)
(True, ok)
(True, ok)
(True, ok)
(True, ok)
(True, ok)
```

**Comment by tannergooding** [comment]
Determinism with -0 vs +0 is actually relevant and expected

The general issue is that the pre-existing CreateReflection01 test is “bad” and is essentially testing the output is the same as base math, which is faulty and not itself “correct”

I’d be fine with leaving the existing test using the helper to minimize PR churn, but the new test should be strictly doing the right thing here

**Comment by hez2010** [comment]
Test failures are unrelated. 

@tannergooding PTAL

**Review by tannergooding — APPROVED** [review]

**Pull request merged**